[UI] Add integration test for sidebar and node page navigation

ChengYanJin opened the pull request:
**Component**: ui, tests

**Context**: 
We want to test the navigation of sidebar and node page

**Summary**:
Add integration test for the sidebar, nodelist and nodetabs in order to test the navigation.

**Acceptance criteria**: 
Pass all the tests.

**Some general questions**: 
- When it come to mock data, what dataset should we use? Should we use the real data from the real platform (6 nodes with *volumes)
- I have the impression that test the stubbed history is too close to the code implementation. Although it's a good approach to avoid the unexpected error from the new page since we only want to check if the URL is correct.

ChengYanJin committed:
ui/ducks/layout: Fix the typo of sidebar_expanded

ChengYanJin committed:
ui/layout: Add data-cy for sidebar in order to do the interaction test

ChengYanJin committed:
ui/cypress: Add integration test for sidebar

ChengYanJin committed:
ui/tests: Add integration test for nodelist and nodetab

ChengYanJin committed:
ui/tests: Add baseURL in cypress.json

ChengYanJin committed:
ui/tests: Add the mock data in fixtures

gdemonet commented on ui/cypress.json:
```diff
@@ -15,5 +15,6 @@
   },
-  "chromeWebSecurity": false
+  "chromeWebSecurity": false,
+  "baseUrl": "http://localhost:3000"
```
Maybe we can get rid of `env.target_url` then?

ChengYanJin replied:
Well, I am not sure we want to remove it. The cost of removing it would be hardcode this `cy.visit(http://localhost:3000)` all the times.

gdemonet replied:
?? Isn't the point of `baseUrl` to always prefix any path for `cy.visit`?
We would just need to replace any `cy.visit(Cypress.env('target_url'))` by `cy.visit('/')`.

ChengYanJin replied:
Indeed!! ;)

gdemonet commented on ui/src/ducks/app/layout.test.js:
```diff
@@ -1,32 +1,32 @@
 
 it('should initToggleSideBar', () => {
-  localStorage.setItem(SIDEBAR_EXPENDED, 'true');
+  localStorage.setItem(SIDEBAR_EXPANDED, 'true');
```
Would make sense to squash this with the typo fix

gdemonet commented on ui/cypress/integration/node/nodelist.spec.js:
```diff
@@ -0,0 +1,34 @@
+// Navigation tests
+describe('Node list', () => {
+  beforeEach(() => {
+    cy.server();
```
No need to re-run `cy.server`, that's already part of `setupMocks`

gdemonet commented on ui/cypress/integration/node/nodelist.spec.js:
```diff
@@ -0,0 +1,34 @@
+describe('Node list', () => {
+  beforeEach(() => {
+    cy.server();
+    cy.route('api/kubernetes/api/v1/nodes', 'fixture:kubernetes/nodes.json');
```
If this route is expected to be useful _in general_, then I'd suggest you define it in `cy.setupMocks`

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,121 @@
+
+  const tabs = ['alerts', 'metrics', 'volumes', 'pods', 'details'];
+  tabs.map((tab) => {
+    it(`brings me to the [${tab}] tab`, () => {
```
We shouldn't use square brackets in test names - if the test fails, the generated screenshot will have square brackets in its name and we cannot upload such files to artifacts today (discovered this when debugging CI issues).

ChengYanJin replied:
Thanks for letting me know this point!

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,121 @@
+});
+
+// Navigation tests
+describe('Overview tab in master-0 node page', () => {
```
That's not a generic "context" name - what you are testing is rather `Node page overview tab` (no need to specify a node name)

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,121 @@
+  const tabs = ['alerts', 'metrics', 'volumes', 'pods', 'details'];
+  tabs.map((tab) => {
+    it(`brings me to the [${tab}] tab`, () => {
+      cy.get('.sc-tabs-item')
+        .eq(tabs.indexOf(tab) + 1)
```
Can we define a `data-cy` attribute for those tabs?

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,121 @@
+      });
+  });
+
+  const tabs = ['alerts', 'metrics', 'volumes', 'pods', 'details'];
+  tabs.map((tab) => {
```
This is not specific to the `overview` tab, so maybe it makes sense get this test in a more generic context (e.g. `Node page tabs`), and make sure that we can reach any tab from any other tab (and also check that clicking the current tab doesn't do anything).

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,121 @@
+  beforeEach(() => {
+    cy.server();
+    cy.route('api/kubernetes/api/v1/nodes', 'fixture:kubernetes/nodes.json');
+    cy.route(
+      'api/prometheus/api/v1/query?query=node_uname_info',
+      'fixture:prometheus/node-uname-info.json',
+    );
```
Same, this one makes sense to keep in the general `setupMocks`

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,121 @@
+  beforeEach(() => {
+    cy.server();
+    cy.route('api/kubernetes/api/v1/nodes', 'fixture:kubernetes/nodes.json');
+    cy.route(
+      '/api/kubernetes/apis/storage.metalk8s.scality.com/v1alpha1/volumes',
+      'fixture:kubernetes/volumes.json',
+    );
```
Ditto

gdemonet reviewed (commented):
Well done :+1: !

> When it come to mock data, what dataset should we use? Should we use the real data from the real platform (6 nodes with *volumes)

Yeah, let's try to get real-world data (hard-coded for now, we'll see how to get them from a real deployment, maybe from CI, to make sure our UI keeps working when APIs evolve).

> I have the impression that test the stubbed history is too close to the code implementation. Although it's a good approach to avoid the unexpected error from the new page since we only want to check if the URL is correct.

What would you rather do? IMHO, all we want is to make sure the "link" is correct. But unlike some elements which have an `href` we can check, most of our "links" are JS-only, so we need to catch the call made to `history.push` to see what their `href` would be.

ChengYanJin committed:
ui/cypress: Remove the env.target_url

Since we add baseUrl http://localhost:3000 in the cypress.json
we don't need the target_url anymore

ChengYanJin committed:
ui/tests: Add data-cy for the components for the tests

ChengYanJin committed:
ui/tests: Add general cy.route in cy.setupMocks()

gdemonet commented on ui/cypress/integration/node/nodelist.spec.js:
```diff
@@ -0,0 +1,64 @@
+    cy.get('[data-cy="Last 7 days"]').click();
+    cy.get('@historyPush').should('be.calledWith', {
+      search: 'from=now-7d',
+    });
```
That's likely not what you want. Here, since you're stubbing `history.push`, you won't have the right URL before switching to `master-1`. That's why your call doesn't contain the `search` part you're expecting.

Instead, you should directly call `cy.visit('/nodes/master-0/metrics?from=now-7d')`.

Also, side-note but quite important: a stub can store many calls for you to check after. When using `should('be.calledWith', args)`, you're only asserting it was called at least once with `args`. In your case, the call with `{ search: 'from=now-7d' }` comes from the click on `cy.get('[data-cy="Last 7 days"]')`, not from the switch to another node.

If we want to clean this test and make sure it does exactly what we want to, here's how we could do it:

```javascript
cy.visit('/nodes/master-0/metrics?from=now-7d');
cy.stubHistory(); // would be smart to create this command, since we seem to use it often
cy.contains('[data-cy="node_table_name_cell"]', 'master-1').click();
cy.get('@historyPush')
  .should('be.calledOnce')
  .and('be.calledWithExactly', {
    pathname: '/nodes/master-1/metrics',
    search: 'from=now-7d',
  });
```

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,173 @@
+  it('does not switch the tab', () => {
+    cy.get(`[data-cy="overview_tab_node_page"]`).click();
+    cy.get('@historyPush').should('not.be.called');
+  });
```
Could be factorized between source tabs:

```javascript
const allTabs = ['overview', 'alerts', 'metrics', 'volumes', 'pods', 'details'];
const tabSwitchingTestCases = (sourceTab) => {
  allTabs.filter((tab) => tab !== sourceTab).map((tab) => {
    it(`brings me to the ${tab} tab`, () => {
      cy.get(`[data-cy="${tab}_tab_node_page"]`).click();
      cy.get('@historyPush').should('be.calledWith', `/nodes/master-0/${tab}`);
    });
  });

  it('does not switch to the current tab', () => {
    cy.get(`[data-cy="${sourceTab}_tab_node_page"]`).click();
    cy.get('@historyPush').should('not.be.called');
  });
};

describe('Overview tab', () => {
  // beforeEach...

  tabSwitchingTestCases('overview');

  // more tests...
});

// more tabs...
```

gdemonet commented on ui/cypress/integration/node/nodetabs.spec.js:
```diff
@@ -0,0 +1,173 @@
+
+  const alertSeverities = ['critical', 'warning'];
+  alertSeverities.map((severity) => {
+    it(`brings me to the alert tab prefiltered by ${severity} severity of master-0 node`, () => {
```
```suggestion
    it(`brings me to the alert tab prefiltered by ${severity} severity`, () => {
```

gdemonet commented on ui/cypress/integration/sidebar.spec.js:
```diff
@@ -0,0 +1,78 @@
+      .its('__history__')
+      .then((history) => {
+        cy.stub(history, 'push').as('historyPush');
+      });
```
Let's at least put this in `beforeEach` :slightly_smiling_face:

gdemonet commented on ui/cypress/integration/sidebar.spec.js:
```diff
@@ -0,0 +1,78 @@
+    cy.get('@historyPush').should('be.calledWith', '/environments');
+  });
+
+  it('can expand', () => {
+    let isSidebarExpanded;
+
+    cy.get('.sc-sidebar div:first .sc-button').click();
+    cy.window()
+      .its('localStorage')
+      .then((store) => {
+        isSidebarExpanded = store.getItem('sidebar_expanded');
+      })
+      .then(() => {
+        // Cypress commands are asynchronous, so you cannot check a property value before the Cypress commands ran.
+        // use cy.then() callback to check the value.
+        expect(isSidebarExpanded).to.equal('false');
+      });
+  });
```
If this test checks that the sidebar `can expand`, I would expect `isSidebarExpanded` to be true, at the end.

Now, this test should not depend on "the default state of the app" (here, it's expanded by default), otherwise the test will fail once we change this default, while the feature still works.

Here's how I would rewrite those tests:

```javascript
  it('is expanded by default', () => {
    // assuming we have set the proper attributes
    // 'data-cy' and 'data-cy-state' in `sidebarConfig`
    cy.get('[data-cy="sidebar"]').should(
      'have.attr',
      'data-cy-state',
      'expanded',
    );

    cy.window()
      .its('localStorage')
      .invoke('getItem', 'sidebar_expanded')
      .should('equal', 'true');
  });

  it('can be collapsed', () => {
    cy.window()
      .its('localStorage')
      .invoke('setItem', 'sidebar_expanded', 'true');

    cy.get('[data-cy="sidebar"] > :first').click();

    cy.get('[data-cy="sidebar"]').should(
      'have.attr',
      'data-cy-state',
      'collapsed',
    );
    cy.window()
      .its('localStorage')
      .invoke('getItem', 'sidebar_expanded')
      .should('equal', 'false');
  });

  it('can be expanded', () => {
    cy.window()
      .its('localStorage')
      .invoke('setItem', 'sidebar_expanded', 'false');

    cy.get('[data-cy="sidebar"] > :first').click();

    cy.get('[data-cy="sidebar"]').should(
      'have.attr',
      'data-cy-state',
      'expanded',
    );
    cy.window()
      .its('localStorage')
      .invoke('getItem', 'sidebar_expanded')
      .should('equal', 'true');
  });

ChengYanJin replied:
In the third case, we may not be able to get the default sidebar_expanded item in` localStorage` though.
Because if do nothing to the sidebar, we don't set it in localStorage. the default state would be in the redux store. 

We could check the data in the redux store, but I think to check the attr of .sc-sidebar the sidebar would be sufficient. Would you think so?

gdemonet replied:
I guess you mean "in the first case", i.e. `it('is expanded by default')`?
I did see this when running the app with a clear localStorage (nothing set for `sidebar_expanded`), but somehow when running using Cypress it was set (and checking that it wasn't set failed...). Any idea why that would be the case?

gdemonet reviewed (changes requested):
Few improvements to make on the tests, but that's starting to look really good! Making changes to core-ui or some navigation logic in the future should be done without worry that it breaks the user-facing behaviour of our UI :ok_hand: :sunglasses:!

ChengYanJin committed:
ui/tests: Create stubHistory command

ChengYanJin commented:
> Few improvements to make on the tests, but that's starting to look really good! Making changes to core-ui or some navigation logic in the future should be done without worry that it breaks the user-facing behaviour of our UI 👌 😎!

Indeed, we do spend some effort on implementing the navigation test, although we can test it manually for the moment. I would agree it's more for the future, our coming changes won't break the UI. 

I also read more about cy.stub(), have a better understanding now^^

Thanks for the review!

ChengYanJin committed:
ui/test: Refactorize the code between source tabs

ChengYanJin committed:
ui/tests: Improve the sidebar integration test

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/add-integration-test-for-sidebar-and-node-page-navigation` into
`development/2.6`
* [w/2.7/feature/add-integration-test-for-sidebar-and-node-page-navigation](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/feature/add-integration-test-for-sidebar-and-node-page-navigation) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# History mismatch



Merge commit #c0c24bc6d1ce7a6c8bc1b6d9ea9fb6f214c67a68 on the integration branch
`w/2.7/feature/add-integration-test-for-sidebar-and-node-page-navigation` is merging a branch which is neither the current
branch `feature/add-integration-test-for-sidebar-and-node-page-navigation` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `feature/add-integration-test-for-sidebar-and-node-page-navigation` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin committed:
ui/tests: Add window.gc()

ChengYanJin committed:
ui/tests: Add the viewportWidth and viewportHeight in cypress.json

In order to configure the browser to have the same viewport as real run application

ChengYanJin committed:
ui/tests: To make sure state is ready for the test

ChengYanJin committed:
ui/tests: Increase the default timeout

ChengYanJin committed:
ui/ducks/monitoring: Make sure the props form the query are ready before sending the request

During the tests, we discovered that there are a lot of query requests were sent
without waiting the props that form the queries ready.

Because of these, huge amount of requests were sent too soon slow down the test.

ChengYanJin committed:
ui/tests: Stub the history after click on the tab

Since we implement the sorting for table, we have `?sort=health` in the history

ChengYanJin committed:
ui/tests: Add more interaction test for nodelist

ChengYanJin committed:
ui/tests: Add/Rename data-cy in the component

ChengYanJin committed:
ui/tests: Add more interaction test for nodetab

ChengYanJin committed:
ui/test: Adapt the test with the new UI changes
